pkg/operator/etcdmemberlistercontroller/etcdmemberlistercontroller.go

```diff
@@ -0,0 +1,86 @@
+package etcdmemberlistercontroller
+
+import (
+	"context"
+	"time"
+
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+	errorsutil "k8s.io/apimachinery/pkg/util/errors"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/cluster-etcd-operator/pkg/dnshelpers"
+	"github.com/openshift/cluster-etcd-operator/pkg/etcdcli"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+// EtcdMemberListerController keeps etcd members cached, and updates the cache when the membership changes
+type EtcdMemberListerController struct {
+	operatorClient v1helpers.OperatorClient
+	etcdClient     etcdcli.EtcdClient
+	membersCache   map[string]*etcdserverpb.Member
+}
+
+func NewEtcdMemberListerController(operatorClient v1helpers.OperatorClient,
+	etcdClient etcdcli.EtcdClient,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	c := &EtcdMemberListerController{
+		operatorClient: operatorClient,
+		etcdClient:     etcdClient,
+		membersCache:   make(map[string]*etcdserverpb.Member),
+	}
+	return factory.New().ResyncEvery(30*time.Second).WithSync(c.sync).ToController("EtcdMemberListerController", eventRecorder.WithComponentSuffix("etcd-member-lister-controller"))
+}
+
+func (c *EtcdMemberListerController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	err := c.syncEtcdMembers(ctx, syncCtx.Recorder())
+	if err != nil {
+		_, _, updateErr := v1helpers.UpdateStatus(ctx, c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "EtcdMemberListerControllerDegraded",
```

Comment:
we should only degrade if this fails for several consecutive times, a transient failure every 30s shouldn't be a big issue to degrade the whole operator on